### PR TITLE
[BEAM-2005] Swap to use Lists within MatchResult instead of arrays.

### PR DESCRIPTION
**Pull request by lukecwik**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
Everywhere in our codebase we go through and want to have lists as the return type. This also AutoValue's them.

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @davorbonaci @dhalperi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java [hunk 1]
```diff
@@ -180,12 +181,12 @@ private MatchResult matchOne(String spec) throws IOException {
     File file = Paths.get(spec).toFile();
 
     if (file.exists()) {
-      return MatchResult.create(Status.OK, new Metadata[]{toMetadata(file)});
+      return MatchResult.create(Status.OK, ImmutableList.of(toMetadata(file)));
     }
 
     File parent = file.getAbsoluteFile().getParentFile();
     if (!parent.exists()) {
-      return MatchResult.create(Status.NOT_FOUND, EMPTY_METADATA);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java]: delete the variable too?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/NumberedShardedFile.java [hunk 2]
```diff
@@ -120,8 +119,8 @@ public String getFilePattern() {
     do {
       try {
         // Match inputPath which may contains glob
-        Collection<Metadata> files = Arrays.asList(Iterables.getOnlyElement(
-            FileSystems.match(ImmutableList.of(filePattern))).metadata());
+        Collection<Metadata> files = Iterables.getOnlyElement(
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/NumberedShardedFile.java]: `List` instead of `Collection`?

Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/NumberedShardedFile.java]: Going to leave as is.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java (hunk 1 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java]: what variable?

Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java]: Figured it out

**Pull request closed without merging**